geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/QueryInterceptor.scala

```diff
@@ -160,14 +164,135 @@ object QueryInterceptor extends LazyLogging {
 
     override def close(): Unit = {}
 
-    private def duration(values: Seq[Bounds[ZonedDateTime]]): FiniteDuration = {
-      values.foldLeft(Duration.Zero) { (sum, bounds) =>
-        sum + Duration(bounds.upper.value.get.toEpochSecond - bounds.lower.value.get.toEpochSecond, TimeUnit.SECONDS)
-      }
-    }
   }
 
   object TemporalQueryGuard {
     val Config = "geomesa.filter.max.duration"
   }
+
+  object GraduatedQueryGuard extends LazyLogging {
+    val ConfigPath = "geomesa.guards.graduated-guard"
+
+    /**
+     *
+     * @param sizeLimit in square degrees
+     * @param durationLimit Maximum duration for a query at or below the spatial size
+     */
+    case class SizeAndDuration(sizeLimit: Int, durationLimit: Duration)
+
+    def loadConfiguration(sft: SimpleFeatureType): Seq[SizeAndDuration]  = {
+      val conf = ConfigFactory.load()
+      try {
+        val guardConfig = conf.getConfigList(s"${ConfigPath}.${sft.getTypeName}")
+        buildLimits(guardConfig)
+      } catch {
+        case e: ConfigException =>
+          logger.error(s"Configuration for GraduatedQueryGuard not available or correct for type ${sft.getTypeName}.  " +
+            s"This query guard is disabled.")
+          Seq()
+      }
+    }
+
+    /**
+     * This function checks conditions on the limits.
+     * 1. Sizes must not be repeated.
+     * 2. Smaller sizes must allow for longer durations.
+     * @param limits Sequence of limits
+     * @return Sorted list of limits or throws an IllegalArgumentException
+     */
+    def evaluateLimits(limits: Seq[SizeAndDuration]): Seq[SizeAndDuration] = {
+      val candidate: Seq[SizeAndDuration] = limits.sortBy(_.sizeLimit)
+      candidate.sliding(2).foreach { seq =>
+        val first = seq.head
+        val second = seq.tail.head
```

Comment:
might be slightly nicer to pattern match like `.foreach { case Seq(first, second) =>`
one other gotcha is that if the list only has one element, `sliding(2)` will give you a list with one element in it...

Reply:
I like the pattern, and was thinking about the first time through; just didn't bother to figure out the syntax:).  I've added a little size check to handle the one limit case.  Maybe that case is degenerate and should throw some kind of error...  Thoughts?

Reply:
i guess you were kind of assuming that there would be a config for the upper limit whole world bounding box, but it doesn't seem required... that probably breaks the logic in the query guard, as it would return None and let the query through (where you probably would expect it to fail?). so we should probably enforce something in the initialization or modify the logic in the guard.